ta/remoteproc/elf_parser.c

```diff
@@ -0,0 +1,174 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2023, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <elf_parser.h>
+#include <string.h>
+#include <trace.h>
+#include <types_ext.h>
+#include <util.h>
+
+static bool va_in_fwm_image_range(void *va, uint8_t *fw, size_t fw_size)
+{
+	uint8_t *vaddr = va;
+
+	return vaddr >= fw && vaddr < fw + fw_size;
+}
+
+TEE_Result e32_parse_ehdr(uint8_t *fw, size_t size)
+{
+	Elf32_Ehdr *ehdr = (Elf32_Ehdr *)(void *)fw;
+
+	if (!fw || !IS_ALIGNED_WITH_TYPE(fw, uint32_t)) {
+		EMSG("Invalid fw address %p", fw);
```

Comment:
I could not fine definitive specification for `PRIxVA`.
I would say `PRIxVA` is rather related to `vaddr_t` (and `PRIxPTR` for `uintptr_t` despite those 2 PRIxXXX are the same).
Recent toolchain seems to accept `%p` for any pointer (`sometype *`) . I remember old toolchains warned on non-`void *` type references.

Reply:
> I could not fine definitive specification for `PRIxVA`. I would say `PRIxVA` is rather related to `vaddr_t`

Right.

> (and `PRIxPTR` for `uintptr_t` despite those 2 PRIxXXX are the same). Recent toolchain seems to accept `%p` for any pointer (`sometype *`) . I remember old toolchains warned on non-`void *` type references.

Then perhaps we should always use `"0x" PRIxPTR` to print pointers?

Also, this is an EMSG(), not a debug message, so I think we should not print an address in clear as it could help an attacker defeat ASLR. Perhaps we should consider doing like in the Linux kernel and have a CFG to hide addresses by default?

Reply:
> Perhaps we should consider doing like in the Linux kernel and have a CFG to hide addresses by default?

Sounds a good idea. I guess this would be out of the scope of this P-R.
In the meantime, @arnopo, maybe the simplest thing here is to not print the virtual address, if you agree.

Reply:
I change the print to `EMSG("Invalid firmware address");`

Reply:
> Then perhaps we should always use "0x" PRIxPTR to print pointers?

I'd rather keep it apart. `PRIxPA` for `paddr_t`, `PRIxVA` for `vaddr_t`, `PRIxPTR` for `uintptr_t`, and `%p` for `void *`.

Reply:
> > Then perhaps we should always use "0x" PRIxPTR to print pointers?
> 
> I'd rather keep it apart. `PRIxPA` for `paddr_t`, `PRIxVA` for `vaddr_t`, `PRIxPTR` for `uintptr_t`, and `%p` for `void *`.

Thanks for clarifying. Indeed the `I` in `PRI` stands for integer so it totally makes sense.